### PR TITLE
[Runtime] Handle substitution when mapping a mangled name to type metadata

### PR DESCRIPTION
**Pull request by DougGregor**

### CODE REVIEW

**Comment by DougGregor** [comment]
@xedin you might be interested in this

File: stdlib/public/core/Misc.swift [hunk 1]
```diff
@@ -102,19 +102,34 @@ func _typeByName(_ name: String) -> Any.Type? {
...
 
 /// Lookup a class given a mangled name. This is a placeholder while we bring
 /// up this functionality.
 public  // TEMPORARY
-func _typeByMangledName(_ name: String) -> Any.Type? {
+func _typeByMangledName(_ name: String,
+                        substitutions: [[Any.Type]] = []) -> Any.Type? {
+  // Map the substitutions to a flat representation that's easier to thread
+  // through to the runtime.
+  let numberOfLevels = UInt(substitutions.count)
```
Comment (slavapestov) [inline stdlib/public/core/Misc.swift]: Why not have the function take the flat representation instead of performing a representation change?

Comment (DougGregor) [reply stdlib/public/core/Misc.swift]: TBH, I'm expecting this testing-only function to get completely removed in the near term, so I don't care much about the API. We'll get to test this code path through the various other adopters soon-ish.

**Comment by DougGregor** [comment]
@swift-ci please test and merge

**Comment by DougGregor** [comment]
@swift-ci please test and merge

**Comment by DougGregor** [comment]
@swift-ci please test and merge

**Comment by DougGregor** [comment]
@swift-ci please test and merge

**Comment by DougGregor** [comment]
@swift-ci please test and merge

**Pull request merged**